inc/sitemaps/class-sitemaps.php

```diff
@@ -0,0 +1,436 @@
+	 */
+	public function reduce_query_load() {
+
+		if ( ! isset( $_SERVER['REQUEST_URI'] ) ) {
+			return;
+		}
+
+		$request_uri = $_SERVER['REQUEST_URI'];
+		$extension   = substr( $request_uri, -4 );
+
+		if ( false !== stripos( $request_uri, 'sitemap' ) && in_array( $extension, array( '.xml', '.xsl' ) ) ) {
+			remove_all_actions( 'widgets_init' );
+		}
+	}
+
+	/**
+	 * Register your own sitemap. Call this during 'init'.
+	 *
+	 * @param string   $name     The name of the sitemap.
+	 * @param callback $function Function to build your sitemap.
+	 * @param string   $rewrite  Optional. Regular expression to match your sitemap with.
+	 */
+	public function register_sitemap( $name, $function, $rewrite = '' ) {
+		add_action( 'wpseo_do_sitemap_' . $name, $function );
+		if ( ! empty( $rewrite ) ) {
+			add_rewrite_rule( $rewrite, 'index.php?sitemap=' . $name, 'top' );
+		}
+	}
+
+	/**
+	 * Register your own XSL file. Call this during 'init'.
+	 *
+	 * @param string   $name     The name of the XSL file.
+	 * @param callback $function Function to build your XSL file.
+	 * @param string   $rewrite  Optional. Regular expression to match your sitemap with.
+	 */
+	public function register_xsl( $name, $function, $rewrite = '' ) {
+		add_action( 'wpseo_xsl_' . $name, $function );
+		if ( ! empty( $rewrite ) ) {
+			add_rewrite_rule( $rewrite, 'index.php?xsl=' . $name, 'top' );
+		}
+	}
+
+	/**
+	 * Set the sitemap current page to allow creating partial sitemaps with wp-cli
+	 * in an one-off process.
+	 *
+	 * @param integer $current_page The part that should be generated.
+	 */
+	public function set_n( $current_page ) {
+		if ( is_scalar( $current_page ) && intval( $current_page ) > 0 ) {
+			$this->current_page = intval( $current_page );
+		}
+	}
+
+	/**
+	 * Set the sitemap content to display after you have generated it.
+	 *
+	 * @param string $sitemap The generated sitemap to output.
+	 */
+	public function set_sitemap( $sitemap ) {
+		$this->sitemap = $sitemap;
+	}
+
+	/**
+	 * Set a custom stylesheet for this sitemap. Set to empty to just remove the default stylesheet.
+	 *
+	 * @param string $stylesheet Full xml-stylesheet declaration.
+	 */
+	public function set_stylesheet( $stylesheet ) {
+		$this->stylesheet = $stylesheet;
+	}
+
+	/**
+	 * Set as true to make the request 404. Used stop the display of empty sitemaps or invalid requests.
+	 *
+	 * @param bool $bool Is this a bad request. True or false.
+	 */
+	public function set_bad_sitemap( $bool ) {
+		$this->bad_sitemap = (bool) $bool;
+	}
+
+	/**
+	 * Prevent stupid plugins from running shutdown scripts when we're obviously not outputting HTML.
+	 *
+	 * @since 1.4.16
+	 */
+	public function sitemap_close() {
+		remove_all_actions( 'wp_footer' );
+		die();
+	}
+
+	/**
+	 * Hijack requests for potential sitemaps and XSL files.
+	 *
+	 * @param \WP_Query $query Main query instance.
+	 */
+	public function redirect( $query ) {
```

Comment:
Yeah, I want to tear out caching into separate class so left this alone before then.